File: src/main/java/org/veupathdb/lib/blast/field/BlastPTask.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.DefaultingJSONValue;
import org.veupathdb.lib.blast.util.JSONConstructor;

public enum BlastPTask implements DefaultingJSONValue
{
  BlastP("blastp"),
  BlastPFast("blastp-fast"),
  BlastPShort("blastp-short");

  public static final BlastPTask DefaultTask = BlastP;

  private final String value;

  BlastPTask(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  @Override
  public boolean isDefault() {
    return this == DefaultTask;
  }

  @JsonValue
  @Override
  public JsonNode toJSON() {
    return JSONConstructor.newText(value);
  }

  public static BlastPTask fromString(String value) {
    for (var val : values())
      if (val.getValue().equals(value))
        return val;

    throw new IllegalArgumentException();
  }

  @JsonCreator
  public static BlastPTask fromJSON(JsonNode js) {
    return fromString(js.textValue());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/BlastPTask.kt
package org.veupathdb.lib.blast.field

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import org.veupathdb.lib.blast.util.DefaultingJSONValue
import org.veupathdb.lib.blast.util.JSONConstructor

enum class BlastPTask(val value: String): DefaultingJSONValue
{
  BlastP("blastp"),
  BlastPFast("blastp-fast"),
  BlastPShort("blastp-short");

  override val isDefault get() = this == DefaultTask

  @JsonValue
  override fun toJSON(): JsonNode {
    return JSONConstructor.newText(value)
  }

  companion object {
    @JvmStatic
    val DefaultTask = BlastP

    @JvmStatic
    fun fromString(value: String): BlastPTask {
      for (tmp in values())
        if (tmp.value == value)
          return tmp

      throw IllegalArgumentException()
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): BlastPTask {
      return fromString(js.textValue() ?: throw IllegalArgumentException())
    }
  }
}
